### PR TITLE
Added ORDER BY to delete operation + update composer.json + fix tests

### PR DESCRIPTION
**Pull request by ildyria**
Copy of https://github.com/LycheeOrg/laravel-nestedset/pull/6 to v6 branch (to align with upstream repo with v6).

> Partially fixes https://github.com/LycheeOrg/Lychee/issues/1251
> 
> As MySQL does not comly to ANSI SQL and does not support statement consistency, but only simple row consistency, the order of rows is important when a complete sub-tree is deleted.

### CODE REVIEW

**Comment by nagmat84** [comment]
I haven't checked the details, but did you only migrate #6 on top of the v6 branch? Please keep in mind that we already had to merge a couple of other bug fixes into our copy of the v5 branch, inter alia

- another fix for deletion some month ago
- a fix which uses a proper interface to check whether a model is a nested set node
- some fixes with respect to exception handling

This are only the fixes a do remember. Maybe there are more.

So basically, our v5 branch and the upstream v5 were diverted (not in the git sense, but in the sense that we had to fix a lot which never made it upstream). Have you checked that all these bug fixes are also in "our" v6?

**Comment by ildyria** [comment]
I took their v6, merged into your branch (which is based on our diverged v5) and created a pull request to a copy of their v6.

If you like `git log --one-line`
2022-04-10_1920x1080_15:31:55

So unless I'm wrong, this PR to v6 basically contains all the changes you made to the v5 + #6 + passing tests :)

**Comment by nagmat84** [comment]
Understood. So I still need to find out what the original difference between upstream v5 and v6 is. I am always a little bit conservative with major upgrades.

**Comment by ildyria** [comment]
> Understood. So I still need to find out what the original difference between upstream v5 and v6 is. I am always a little bit conservative with major upgrades.

It is just a bump of the Laravel version, code wise there is pretty much 0 change... I seriously do not know why they decided to make it v6. IMHO that was not needed.
The only reason why I would be in favor to align with them is to make it easier to integrate some of the potential fixes they may have in the future.

You can clearly see that their v6 does not add anything, especially given that the merge commit is quite empty:
https://github.com/LycheeOrg/laravel-nestedset/pull/7/commits/f47602efb00b58d68b1f3f8a3ae61561ee4673aa

As for https://github.com/LycheeOrg/laravel-nestedset/pull/7/commits/ec8e1d706187eb672e111b36e4cca8c28229e202
It is a fix that was applied on the v5 after the v5-v6 split on their repo.
I took the initiative to merge it here.

File: src/BaseRelation.php [hunk 1]
```diff
@@ -144,7 +144,7 @@ public function addEagerConstraints(array $models)
         // The first model in the array is always the parent, so add the scope constraints based on that model.
         // @link https://github.com/laravel/framework/pull/25240
         // @link https://github.com/lazychaser/laravel-nestedset/issues/351
-        optional($models[0])->applyNestedSetScope($this->query);
+        optional(reset($models))->applyNestedSetScope($this->query);
```
Comment (nagmat84) [inline src/BaseRelation.php:147]: Why do we change this line? It seems that in upstream v5 it was `optional(reset($models))` and was changed into `optional($models[0])` for upstream v6. So with the PR we basically undo that upstream change from v5 to v6. Why?

File: tests/NodeTest.php [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 class NodeTest extends PHPUnit\Framework\TestCase
 {
-    public static function setUpBeforeClass(): void
+    public static function setUpBeforeClass() : void
```
Comment (nagmat84) [inline tests/NodeTest.php]: Since when do we have a blank before the colon of the return type? This is neither the upstream formatting style nor our formatting style for Lychee. If possible, I would like avoid this change in formatting, because it generates unnecessarily large diffs. (Here and below).

File: tests/NodeTest.php [hunk 3]
```diff
@@ -224,7 +224,6 @@ public function testCategoryMovesUp()
     public function testFailsToInsertIntoChild()
     {
         $this->expectException(Exception::class);
-
```
Comment (nagmat84) [inline tests/NodeTest.php]: Similar comment as before. If possible I would like to not change the formatting. In this case, keep the blank line.

File: tests/NodeTest.php (hunk 2 above)
Comment (ildyria) [reply tests/NodeTest.php]: Oups, I may have F up the merge, because I thought I kept our notation.

File: src/BaseRelation.php (hunk 1 above)
Comment (ildyria) [reply src/BaseRelation.php:147]: https://github.com/lazychaser/laravel-nestedset/pull/552
https://github.com/lazychaser/laravel-nestedset/pull/547

Comment (nagmat84) [reply src/BaseRelation.php:147]: Ah, I see. The change from `$models[0]` to `reset($models)` was a fix which for some reason was only applied to their v5 (in lazychaser#547) but did not yet make it into their v6 (because lazychaser#552). So it is basically a upstream regression bug which you already fixed for "our" v6.

File: tests/NodeTest.php (hunk 3 above)
Comment (ildyria) [reply tests/NodeTest.php]: Fixed.

File: tests/NodeTest.php (hunk 2 above)
Comment (ildyria) [reply tests/NodeTest.php]: fixed.

File: README.markdown [hunk 4]
```diff
@@ -1,22 +1,19 @@
...
-[![License](https://poser.pugx.org/kalnoy/nestedset/license.svg)](https://packagist.org/packages/kalnoy/nestedset)
+[![Total Downloads](https://poser.pugx.org/lychee-org/nestedset/downloads.svg)](https://packagist.org/packages/lychee-org/nestedset)
+[![Latest Stable Version](https://poser.pugx.org/lychee-org/nestedset/v/stable.svg)](https://packagist.org/packages/lychee-org/nestedset)
+[![Latest Unstable Version](https://poser.pugx.org/lychee-org/nestedset/v/unstable.svg)](https://packagist.org/packages/lychee-org/nestedset)
+[![License](https://poser.pugx.org/lychee-org/nestedset/license.svg)](https://packagist.org/packages/lychee-org/nestedset)
 
 This is a Laravel 4-8 package for working with trees in relational databases.
 
+It is a fork of [lazychaser/laravel-nestedset](https://github.com/lazychaser/laravel-nestedset) and contains patches which are required for using the library with [Lychee](https://github.com/LycheeOrg/Lychee).
+
```
Comment (nagmat84) [inline README.markdown]:
```suggestion
It is a fork of [lazychaser/laravel-nestedset](https://github.com/lazychaser/laravel-nestedset) and contains general patches which are required for using the library with [Lychee](https://github.com/LycheeOrg/Lychee). Note that the patches are **not** specific for Lychee, but a generally useful. Inter alia:

 * Routines respect a foreign key constraint on the parent-child-relation by taking care that changes to the tree are applied in the correct order.
 * The code does not fail if the model which uses `NoteTrait` does not directly extend `Model` but indirectly inherits `Model` via another parent class.

```

A little bit longer than your text, but I think it is important to stress that Lychee is not the cause of the probles.

File: composer.json [hunk 5]
```diff
@@ -41,10 +51,5 @@
...
             ]
         }
-    },
-    "scripts": {
-        "test": [
-            "@php ./vendor/bin/phpunit"
-        ]
     }
-}
+}
```
Comment (nagmat84) [inline composer.json]:
```suggestion
}

```

**Review by nagmat84 — COMMENTED** [review]
Some minor changes

**Review by nagmat84 — APPROVED** [review]
LGTM.
